File: 1161-maximum-level-sum-of-a-binary-tree/1161-maximum-level-sum-of-a-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
from collections import deque
class Solution:
    def maxLevelSum(self, root: Optional[TreeNode]) -> int:
        if not root:
            return 0
        
        queue = deque([root])
        level = 0
        maxSum = float('-inf')
        maxLevel = 1


        while queue:
            length = len(queue)
            level += 1
            levelSum = 0

            for i in range(length):
                node = queue.popleft()

                #Append max val
                levelSum += node.val

                #We at last val, so add key = level, value = maxVal
                if i == length - 1:
                    if levelSum > maxSum:
                        maxLevel = level
                    maxSum = max(maxSum, levelSum)
                
                if node.left:
                    queue.append(node.left)
                if node.right:
                    queue.append(node.right)
        
        return maxLevel
                

File: 1161-maximum-level-sum-of-a-binary-tree/1161-maximum-level-sum-of-a-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int maxLevelSum(TreeNode* root) {

        std::queue<TreeNode*> queue;
        int level = 0;
        queue.push(root);

        while (!queue.empty())
        {
            int n = queue.size();
            int count = 0;
            level += 1;

            for (int i = 0; i < n; i++)
            {
                TreeNode* val = queue.front();
                queue.pop();

                //Add to count
                count += val->val;
                
                if (val->left)
                {
                    queue.push(val->left);
                }

                if (val->right)
                {
                    queue.push(val->right);
                }
            }

            if (count > res)
            {
                res = count;
                resLevel = level; 
            }
        }

        return resLevel;
    }

private:
    int res = INT_MIN;
    int resLevel = INT_MIN;
};